include/dca/math/nfft/dnfft_1d_gpu.hpp

```diff
@@ -0,0 +1,245 @@
+//
+// Author: Giovanni Balduzzi (gbalduzz@gitp.phys.ethz.ch)
+//
+// This class implements the 1D delayed-NFFT (d-NFFT) algorithm on the GPU.
+// See dnfft_1d.hpp for references and CPU implementation.
+
+#ifndef DCA_HAVE_CUDA
+#error "This file requires CUDA."
+#endif
+
+#ifndef DCA_MATH_NFFT_DNFFT_1D_GPU_HPP
+#define DCA_MATH_NFFT_DNFFT_1D_GPU_HPP
+
+#include "dca/math/nfft/dnfft_1d.hpp"
+
+#include <cuda_runtime.h>
+
+#include "dca/linalg/matrix.hpp"
+#include "dca/linalg/util/allocators.hpp"
+#include "dca/linalg/util/cuda_event.hpp"
+#include "dca/linalg/vector.hpp"
+#include "dca/math/nfft/kernels_interface.hpp"
+#include "dca/phys/domains/quantum/electron_band_domain.hpp"
+
+namespace dca {
+namespace math {
+namespace nfft {
+// dca::math::nfft::
+
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling = 8, NfftModeNames mode = CUBIC>
+class Dnfft1DGpu;
+
+// Only the CUBIC mode is implemented.
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling>
+class Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling, CUBIC>
+    : public Dnfft1D<ScalarType, WDmn,
+                     func::dmn_variadic<func::dmn_0<phys::domains::electron_band_domain>,
+                                        func::dmn_0<phys::domains::electron_band_domain>, RDmn>,
+                     oversampling, CUBIC> {
+public:
+  using BDmn = func::dmn_0<phys::domains::electron_band_domain>;
+  using PDmn = func::dmn_variadic<BDmn, BDmn, RDmn>;
+
+  using ThisType = Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling>;
+  using ElementType = ScalarType;
+  using BaseClass = Dnfft1D<ScalarType, WDmn, PDmn, oversampling, CUBIC>;
+
+public:
+  Dnfft1DGpu(double beta, cudaStream_t stream, bool accumulate_m_sqr = false);
+
+  // Resets the accumulated quantities. To be called before each DCA iteration.
+  void initialize();
+
+  // Accumulates asynchronously on the device the entries of the M matrix at times and orbitals
+  // described by the provided configuration.
+  // Postcondition: M and config shall not be modified untill 'synchronizeCopy' is called.
+  template <class Configuration>
+  void accumulate(const linalg::Matrix<double, linalg::CPU>& M, const Configuration& config,
+                  const int sign);
+
+  // Transforms the accumulated data to the frequency domain and stores it in f_w.
+  // If get_square is true, transforms the accumulated squared data instead.
+  template <typename OtherScalarType>
+  void finalize(func::function<std::complex<OtherScalarType>, func::dmn_variadic<WDmn, PDmn>>& f_w,
+                bool get_square = false);
+
+  // Sums the accumulated data in the time domain.
+  ThisType& operator+=(ThisType& other);
+
+  // Ensures that the arguments of the method 'accumulate' have been copied to the GPU.
+  void synchronizeCopy() {
+    m_copied_event_.block();
+  }
+
+private:
+  void initializeDeviceCoefficients();
+
+private:
```

Comment:
Again, I don't think a second private is necessary.

Reply:
Changed in 547604f176339765f2a44aac02708f7b0ec5dcb7